### PR TITLE
Add support for symfony/lock

### PR DESCRIPTION
**Pull request by digilist**
This PR adds the ability to use a symfony/lock storage to prevent command overlapping, as discussed in #125.

I did not remove the file based locking, as this might be BC break (because of the lockFile() method etc., which is not private). However, this could be removed with the next major release.

This should also solve the request of @IkeLutra, because he can now use the FlockStorage and define the path.

Additionally, I added a unit test that verifies that the task lockig is working properly.

### CODE REVIEW

**Comment by PabloKowalczyk** [comment]
Hello @digilist, thank you very much! I appreciate your work, but, sorry, you should target branch 1.x, as this is active development branch. Could you please backport your PR to PHP 5.6? If you want you can close this PR and open new.
Thanks again, good work.

**Comment by digilist** [comment]
No problem, I was able to rebase the commits and change the target branch of this PR.

**Comment by PabloKowalczyk** [comment]
Ah yes, very nice. There are some problems with AppVeyor CI, i will monitor it, but fail(s) can be unrelated to your changes.

File: composer.json [hunk 1]
```diff
@@ -36,6 +36,7 @@
         "symfony/console": "^2.7 || ^3.3 || ^4.0",
         "symfony/dependency-injection": "^2.7 || ^3.3 || ^4.0",
         "symfony/filesystem": "^2.7 || ^3.3 || ^4.0",
+        "symfony/lock": "^3.4 || ^4.1",
```
Comment (PabloKowalczyk) [inline composer.json]: Is version `^4.1` required, can this be lowered to `^4.0` to be in line with other Symfony dependencies?

File: tests/Functional/PreventOverlappingTest.php [hunk 2]
```diff
@@ -0,0 +1,130 @@
+<?php
+
+namespace Crunz\Tests\Functional;
```
Comment (PabloKowalczyk) [inline tests/Functional/PreventOverlappingTest.php]: This test seems to be `unit`, but it is in `Functional` namespace, it is intended?

**Comment by PabloKowalczyk** [comment]
Could you also add example, of locking to different path (as mentiond in #125), to README?

File: tests/Functional/PreventOverlappingTest.php (hunk 2 above)
Comment (digilist) [reply tests/Functional/PreventOverlappingTest.php]: You are right. At first I thought it's gonna be more of a functional test, I'll move it.

File: composer.json (hunk 1 above)
Comment (digilist) [reply composer.json]: yep, true.

**Comment by digilist** [comment]
I added some examples to the README.

Regarding AppVeyor I do not see how the failing tests could be related to my changes and do not think it is :confused:

**Comment by PabloKowalczyk** [comment]
Looks like the issue with AppVeyor is related to this PR (last build https://ci.appveyor.com/project/lavary/crunz/builds/19558199/job/5egmyi1aw64uhw08), CI killed the build after 1h, something blocked the end of PHPUnit. On the other hand my PR without `symfony/lock` was successful (https://github.com/lavary/crunz/pull/155).

I have some idea. On Lock component's docs (https://symfony.com/doc/3.4/components/lock.html#flockstore) we can read:

> The FlockStore uses the file system on the local computer to create the locks. It does not support expiration, but the lock is automatically released when the PHP process is terminated

Also some warning:

> Beware that some file systems (such as some types of NFS) do not support locking. In those cases, it's better to use a directory on a local disk drive or a remote store based on Redis or Memcached.

This case reminds me also bug PHPCsFixer https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3460 which was related to file locks on Vagrant's NFS.

So, to the conclusion, i think that weird behavior of AppVeoyr is related to `FlockStore`. For some reason PHP can't release Lock (maybe AppVeoyor uses NFS, IDK) and process hangs for 1h.
IMO this issue can be easy fixed by swapping lock store to `SemaphoreStore`.
WDYT?

**Comment by digilist** [comment]
Interesting idea, thanks. It looks like the file locking in general is the problem. Even the "old" code does not work, it's not only symfony lock. And as there was no test case for the locking previously, it is also no longer surprising for me, that it is failing with my changes now :stuck_out_tongue_winking_eye: 

I only noticed just now, that AppVeyor is based on Windows, so the problem is (probably) related to Windows.

I changed the condition of the loop from `while($event->isLocked())` to `while($event->getProcess()->isRunning())` and the build is no longer timing out. Now, the last assertion fails, because the event is still locked, but I do not yet know why. 

I will continue my testing to find a solution. I will clean up my test commits later - and I am sorry for that many pushes (if you get an email from github for every push ;)). I currently see no other way to debug it, I have no Windows system at hand.

**Comment by digilist** [comment]
Thanks! I found the issue. It was related to my overriden `manageStartedEvents()` function, where I forgot to call the afterCallbacks of the event, after the process finished. So the lock file on Windows was not deleted and the process was locked forever.

I noticed some issue on Windows here: If the lock file is not deleted, because the EventRunner dies (for whatever reason), the event will never start again. This is because the event is considered locked if the lock file exists (https://github.com/lavary/crunz/blob/1.x/src/Event.php#L1028).

I found this solution to check if a process is actually still running: https://stackoverflow.com/questions/13648304/detecting-if-a-windows-process-and-application-is-running. Probably, we can fix it this way, but I would do it in another PR.

**Comment by digilist** [comment]
I would say so, yes.

**Comment by PabloKowalczyk** [comment]
Thanks @digilist!

**Pull request merged**